### ISSUE TITLE
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb0 in position 6124037: invalid start byte

### ISSUE THREAD

**Issue by pschonmann** [yaroslaff/ncdu-compare#1]
Hi im trying to compare two ncdu outs. But still getting error
with command ncdu  -0xo- /data/www/ --exclude /data/www/domains/tmp > blah

Ncdu is this version
ncdu 1.15.1

```
pschonmann@E495:~/temp$ ncdu-compare ncdu_2023-02-09_094201 ncdu_2023-02-09_094301 
Traceback (most recent call last):
  File "/home/pschonmann/.local/bin/ncdu-compare", line 137, in <module>
    main()
  File "/home/pschonmann/.local/bin/ncdu-compare", line 112, in main
    old_data = json.load(fh)
  File "/usr/lib/python3.10/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/lib/python3.10/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb0 in position 6124037: invalid start byte
```

**Comment by yaroslaff** [comment]
Thank you for reporting problem! 
Probably some file has really strange unicode name. It works for me with cyrillic letters in filename, but anyway, I've fixed opening files, now it open files as utf-8 encoding. 
Could you try new version now? (`pip3 install -U ncdu-compare` to upgrade to 0.0.2)

**Comment by pschonmann** [comment]
Thanks for quick response, but fix didnt help. Still getting  same error.

**Comment by yaroslaff** [comment]
Oops, I added option to wrong line :-).
Please try upgrade it again (new version should be 0.0.3) and try again.

**Comment by pschonmann** [comment]
Nope. Still getting error. The char is celsius degree :)

{"name":"2634-excelent-11°.jpg","asize":51839,"dsize":53248,"ino":4989815},

**Comment by pschonmann** [comment]
When i removed from file. another UTF-8 char appears
https://www.fileformat.info/info/unicode/char/00a0/index.htm

Temporarily fixed with

iconv -c -f utf-8 -t ascii ncdu_2023-02-09_103801 > 1.ascii
iconv -c -f utf-8 -t ascii ncdu_2023-02-10_223801 > 2.ascii
ncdu-compare 1.ascii 2.ascii

**Comment by yaroslaff** [comment]
I'm trying, but cannot reproduce problem:
~~~
(p10) xenon@mir:/tmp/ncdu$ echo asdf > asdf.txt
(p10) xenon@mir:/tmp/ncdu$ ncdu -0xo /tmp/1
(p10) xenon@mir:/tmp/ncdu$ echo asdf > 2634-excelent-11°.jpg
(p10) xenon@mir:/tmp/ncdu$ ncdu -0xo /tmp/2
(p10) xenon@mir:/tmp/ncdu$ ncdu-compare /tmp/1 /tmp/2
5 /tmp/ncdu (5 > 10)
5 /tmp/ncdu/2634-excelent-11°.jpg (0 > 5)
(p10) xenon@mir:/tmp/ncdu$ ls
'2634-excelent-11'$'\302\260''.jpg'   asdf.txt
~~~

Could you reliable reproduce it?

**Comment by yaroslaff** [comment]
Also, please make sure you are testing with new version (`ncdu-compare --help` should print: Compare old and new ncdu dumps. ver. 0.0.5)

**Comment by pschonmann** [comment]
Yeap, i have version updated from pip3 command described above and its 0.0.5 where speedup from #2  works flawlessly.

**Comment by yaroslaff** [comment]
Does this problem occurs on other machines? Maybe you can give some instructions how I could reproduce it? 
for me it worked well with Cyrillic filenames and I created `echo asdf > 2634-excelent-11°.jpg` - .

Also, can you try this simple script (jsonload.py):
~~~~python
#!/usr/bin/env python

import json
import sys

with open(sys.argv[1], encoding="utf-8") as fh:
    data = json.load(fh)

print(json.dumps(data, indent=4))
~~~~

Does it loads (./jsonload.py filename.json)  your ncdu snapshot files or crashes?

**Comment by pschonmann** [comment]
TLDR; The problem is in ncdu generated files.

Script and ncdu-compare fails on all my testing computers ( rpi deb, deb11, mint )

When i "convert" original ncdu file with
jq < ncdu_2023-02-09_103801 > ncdu_2023-02-09_103801.json and second file too ncdu-compare works with newly generated jsons fine

Im not sure if jq do anything else with input, but expecting only pretty output...

**Comment by yaroslaff** [comment]
This is interesting... if ncdu produces valid JSON, why python sometimes cannot read... or is it invalid?
Can you try any online JSON validator to validate ncdu files - is it valid JSON or not?

Or maybe you can make small minimal example to reproduce problem, like .tar.gz with small directory with 1-2 files for which ncdu would produce broken JSON? (probably, only filenames are important, with any files content (zeroes, lorem ipsum) or no content at all, so even if you have sensitive/confidential info on your disk, it's not needed for this example)

**Comment by yaroslaff** [comment]
I've found this: https://dev.yorhel.nl/ncdu/jsonfmt

> Miscellaneous notes
> As mentioned above, file/directory names are not converted to any specific encoding when exporting. If you want the exported info dump to be valid JSON (and thus valid UTF-8), you’ll have to ensure that you have either no non-UTF-8 filenames in your filesystem, or you should process the dump through a conversion utility such as iconv.

Could it be that your filenames are non-utf8?

**Comment by pschonmann** [comment]
Here is that tricky boi.
[onedir.txt](https://github.com/yaroslaff/ncdu-compare/files/10737617/onedir.txt)

chardetect says its not utf-8 file :( 
I dont know know how to force ncdu output json in utf8 and probably detect encoding is the right way for this problem

https://stackoverflow.com/a/31625557/14817216

**Comment by yaroslaff** [comment]
Fixed :-) Please upgrade to v 0.0.6 (latest published) and try it. With -e strict/ignore/replace you can select how it will handle unicode errors. Thank you for bug report and help with debugging it!

**Comment by pschonmann** [comment]
Hooray ! It works.
It was a pleasure to work with you.
Have a nice day !
